UM-USElab/gradecraft-development
# 3001 multiple canvas institutions

### Status
READY

### Description
This PR addresses the desire to have alternative institutions outside of UM Ann Arbor. In order to achieve this, we can now store OAuth credentials such as the `consumer_key`, `consumer_secret`, and `base_url` on the newly-added `Provider` model. A `Provider` belongs to an `Institution`, which has many `Courses`.

See Steps to Test or Reproduce for further details regarding this implementation.

### Todos

### Deploy Notes
Includes migrations to edit and add new models

### Gem dependencies
N/A

### Migrations
YES

### Steps to Test or Reproduce
These changes preserve the ability to use a single site by allowing for the setting of ENV variables for `CANVAS_CLIENT_ID`, `CANVAS_CLIENT_SECRET`, and `CANVAS_BASE_URL`. However, for our purposes, these values should be removed and stored on the `Provider` table so that the provider can be dynamically configured in the app.

To do this, go to the Course index page where you will notice the option in the upper right-hand corner to create a new institution. Fill in the required fields, including the OAuth credentials. Go and edit a course in Gradecraft that you would like to link to Canvas and make sure it is linked to the appropriate institution in the Admin Only panel.

The credentials used for Canvas should be pulled from the Provider table based on the current course's relationship with the Institution model.

### Impacted Areas in Application
All current Canvas functionality in Gradecraft including grade and assignment import.

======================
Closes #3001 
Closes #3283
